File: python/boj/stack/4949.py
while True:
    x=input()
    if x == '.':
        break
    stack=[]
    state=True

    for i in x:
        if i == '(' or i == '[':
            stack.append(i)
        elif i == ')':
            if len(stack)==0:
                state=False
                break
            if stack[-1] == "(":
                stack.pop()
            else:
                state = False
                break
        elif i == ']':
            if len(stack)==0:
                state=False
                break
            if stack[-1] == "[":
                stack.pop()
            else:
                state = False
                break
    if state == True and not stack:
        print('yes')
    else:
        print('no')

    

File: cpp/stack/4949.cpp
#include <iostream>
#include <stack>

using namespace std;

int main(){
    while (true){
        string str;
        stack<char> s;
        bool state = true;
        getline(cin, str);
        if (str=="."){
            break;
        }
        for(int i=0;i<str.length();i++){
            if(str[i]=='('){
                s.push('(');
            }else if(str[i]=='['){
                s.push('[');
            }else if(str[i]==')'){
                if(!s.empty()&&s.top()=='('){
                    s.pop();
                }else{
                    cout << "no" << '\n';
                    state = false;
                    break;
                }
            }else if(str[i]==']'){
                if(!s.empty()&&s.top()=='['){
                    s.pop();
                }else{
                    cout << "no" << '\n';
                    state = false;
                    break;
                }
            }
        }
        if (state == false){
            continue;
        }
        if(s.empty()){
            cout << "yes" << '\n';
        }else{
            cout << "no" << '\n';
        }
        

    }
}